src/main/java/org/assertj/core/internal/Iterables.java

```diff
@@ -453,16 +453,62 @@ public void assertContainsOnlyNulls(AssertionInfo info, Iterable<?> actual) {
    * @throws AssertionError if the given {@code Iterable} does not contain the given sequence of objects.
    */
   public void assertContainsSequence(AssertionInfo info, Iterable<?> actual, Object[] sequence) {
-    if (commonCheckThatIterableAssertionSucceeds(info, actual, sequence)) return;
-    // check for elements in values that are missing in actual.
-    List<?> actualAsList = newArrayList(actual);
-    for (int i = 0; i < actualAsList.size(); i++) {
-      // look for given sequence in actual starting from current index (i)
-      if (containsSequenceAtGivenIndex(actualAsList, sequence, i)) return;
+    // perform the checks that would have been done in commonCheckThatIterableAssertionSucceeds
+    // but do them explicitly without having to create a new iterator on actual - which would
+    // break if actual were only singly-traversable.
+    checkNotNullIterables(info, actual, sequence);
+    // store the elements from actual that have been visited (because we don't know we can look ahead - the 'actual'
+    // might be singly-traversable) in a cyclic list, which can then be rolled for each new element
+    // until a match is found or until the 'actual' is exhausted. Of course if 'actual' really is infinite then this could
+    // take a while :-D
+    // once it's full, check if it equals sequence for each new element
+    RingBuffer buffer = new RingBuffer(sequence.length);
+    int howMany=0;
+    final Iterator<?> actual_iterator = actual.iterator();
+    // ... perform the remaining "normal" checks explicitly
+    if(!actual_iterator.hasNext() && sequence.length==0) return;
+    failIfEmptySinceActualIsNotEmpty(sequence);
+    while(actual_iterator.hasNext()) {
+      final Object next = actual_iterator.next();
+      buffer.add(next);
+      howMany++;
```

Comment:
I don't think we need howMany if RingBuffer gives how many elements it contains.
the if condition could be: 
```
if(buffer.size()>=sequence.length && buffer.equals(sequence))
```
We can even go further and let simply use `if(buffer.equals(sequence))` as buffer.equals checks the sizes of the sequence against the buffer's number of elements. I'm not against keeping `buffer.size()>=sequence.length` in the if condition as it makes it clear that there is no point of doing a comparison if buffer does not have enough elements but that's a detail.

Reply:
That's a good spot actually. I definitely think that buffer.equals(sequence) will obviously return false if they are of a different length so let's trim that howMany reference out ...